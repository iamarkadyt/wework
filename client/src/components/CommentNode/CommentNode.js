import React from 'react'
import Field from '../Field/Field'
import { connect } from 'react-redux'
import { deleteComment } from '../../state/actions/postsActions'
import { fetchUsersStats } from '../../state/actions/userActions'

const dateFormatOptions = {
    year: 'numeric',
    month: 'long',
    day: 'numeric',
    hour: 'numeric',
    minute: 'numeric',
    second: 'numeric'
}

const CommentNode = ({
    _id,
    text,
    date,
    user: {
        _id: authorId,
        name
    },
    postId,
    authedUserId,
    deleteComment,
    fetchUsersStats
}) => {
    const belongsToAuthedUser = authorId === authedUserId
    return (
        <div className="CommentsView-CommentNode-container">
            <p><b>{name}</b> said at {new Date(date)
                .toLocaleDateString('en-US', dateFormatOptions)}:</p>
            <p className="CommentsView-CommentNode-container-Node-body">{text}</p>
            {belongsToAuthedUser
                ? <div>
                    <Field
                        type="linkButton"
                        label="Delete"
                        inline
                        onClick={() =>
                            deleteComment(postId, _id, fetchUsersStats)
                        } />
                </div>
                : null}
        </div>
    )
}

export const _UnconnectedCommentNode = CommentNode
export default connect(state => ({
    authedUserId: state.user.id
}), { deleteComment, fetchUsersStats })(CommentNode)
