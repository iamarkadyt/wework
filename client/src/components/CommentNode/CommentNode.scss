.CommentNode-container {
    .CommentNode-container-Node-body {
        font-size: 1.3rem;
    }
    margin-top: 1rem;
}
