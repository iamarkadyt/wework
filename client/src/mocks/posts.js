export const mockPost = {
    "_id": "c717a50ac29bacb2ee",
    "user": {
        "_id": "some_alphanumeric_gibberish",
        "name": "Dwayne Johnson",
        "avatar": ""
    },
    "text": "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.",
    "comments": [
        {
            "user": {
                "_id": "some_alphanumeric_gibberish2",
                "name": "Vin Diesel",
                "avatar": ""
            },
            "text": "Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "date": "1535133201"
        },
        {
            "user": {
                "_id": "some_alphanumeric_gibberish2",
                "name": "Vin Diesel",
                "avatar": ""
            },
            "text": "Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "date": "1535133201"
        },
        {
            "user": {
                "_id": "some_alphanumeric_gibberish2",
                "name": "Vin Diesel",
                "avatar": ""
            },
            "text": "Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.",
            "date": "1535133201"
        }
    ],
    "likes": [
        {
            "_id": "19825918265102965",
            "user": "aoeu9a8o10s392h0"
        },
        {
            "_id": "19825918265102966",
            "user": "aoeu9a8o10s392hx0"
        },
        {
            "_id": "19825918265102967",
            "user": "aoeu9a8o10s392120"
        }
    ],
    "date": "1535133024"
}

export default new Array(10).fill(mockPost, 0, 10)
