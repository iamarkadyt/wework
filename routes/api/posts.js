const express = require('express');
const router = express.Router();

// Can setup middleware like this:
// router.use(functions);

// @route   GET api/posts/test
// @desc    Tests posts route.
// @access  Public
router.get('/test', (req, res) => {
    res.send('OK');
});

module.exports = router;