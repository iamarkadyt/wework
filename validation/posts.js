const Yup = require('yup')

module.exports = profile => new Promise((resolve, reject) => {
    const schema = Yup.object().shape({
        text: Yup.string()
            .min(10, "Minimum of 10 characters is required")
            .max(300, "Maximum of 300 characters is allowed")
            .required("Please enter some text here")
    })

    schema.validate(profile, { abortEarly: false, strict: true })
        .then(profile => resolve(profile))
        .catch(err => {
            const errors = {}
            for (let i of err.inner)
                errors[i.path] = i.message
            reject(errors)
        })
})