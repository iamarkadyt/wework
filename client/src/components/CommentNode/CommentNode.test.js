/**
 * Contract:
 *
 * What do I do with the props I receive? (Where does rendering vary? etc.)
 * - deleteComment function is called when Delete btn is clicked
 * - fetchUserStats is passed to deleteComment
 * - authedUserId is used to check if comment belongs to curr. user which controls render of button line
 * - postId is used in deleteComment()
 * - comment's fields are simply rendered 
 *
 * What components do I render?
 * Timestamp line, text line and buttons line
 *
 * What props do I pass them?
 * - First <p> has to receive [name] and that must be followed by correct instance of Date object
 * - Second <p> has to receive [text]
 * - Third element - div, is rendered conditionally (uses authedUserId and comment id to determine)
 * - and renders a Field with a set of props: onClick with deleteComment and few other props.
 *
 * What happens if user interacts with any of the components?
 * - Clicking on abovementioned Field invokes deleteComment() function
 *
 * What happens if child component calls a callback passed to it?
 * - if deleteComment is called, instance ceases to exist
 *
 * What happens in lifecycle hooks?
 * - N/A, functional component
 *
 * What about state? How do I update it?
 * - N/A, functional component
 *
 * Context effects?
 * - N/A
 *
 * Public ref interface?
 * - N/A
 *
 */

import React from 'react'
import { shallow } from 'enzyme'
import { _UnconnectedCommentNode as CommentNode, mapStateToProps } from './CommentNode'
import cloneDeep from 'lodash.clonedeep'
import { mockPost } from '../../mocks/posts.js'
import moment from 'moment'

describe('CommentNode', () => {
  let props
  let mountedCommentNode

  const commentNode = () => {
    if(!mountedCommentNode) {
      mountedCommentNode = shallow(<CommentNode {...props} />)
    }
    return mountedCommentNode
  }

  const getMockProps = () => {
    const actions = {
      deleteComment: jest.fn(),
      fetchUsersStats: jest.fn()
    }

    const comment = mockPost.comments[0]

    return cloneDeep({
      ...actions,
      comment,
      postId: mockPost._id,
      authedUserId: mockPost.comments[0].user._id
    })
  }

  beforeEach(() => {
    props = getMockProps()
    mountedCommentNode = undefined
  })

  describe('conditional rendering', () => {
    it('buttons line is shown if user is comment author', () => {
      const authedUserId = 'some_authed_user_id123'
      props.authedUserId = authedUserId
      props.comment.user._id = authedUserId
      expect(commentNode().find({ type: 'linkButton' }).length).toBeGreaterThan(0)
    })

    it('buttons line is not shown if user is not comment author', () => {
      const authedUserId = 'some_authed_user_id123'
      const commenterId = 'some_commenter_id123'
      props.authedUserId = authedUserId
      props.comment.user._id = commenterId
      expect(commentNode().find({ type: 'linkButton' }).length).toBe(0)
    })
  })

  describe('interaction', () => {
    it('deleteComment() is called when Delete button is clicked', () => {
      props.deleteComment = jest.fn()
      commentNode().find({ type: 'linkButton', label: 'Delete' }).simulate('click')
      expect(props.deleteComment).toHaveBeenCalled()
    })
  })

  describe('props/args passing', () => {
    it('timestamp line displays commenter name', () => {
      const name = 'Plain Jane'
      props.comment.user.name = name
      expect(commentNode().find("p").first().text()).toMatch(name)
    })

    it('timestamp line displays correct timestamp', () => {
      const text = commentNode().find(".CommentNode-container").childAt(0).text()
      const substr = text.substring(text.indexOf('at ') + 3)
      // Example: November 30, 2018, 3:18:44 PM:
      expect(moment(substr, 'MMMM D, YYYY, h:mm:ss A:', true).isValid()).toBe(true)
    })
    
    it('text line displays text', () => {
      const text = 'test_text_texst_12345'
      props.comment.text = text
      expect(commentNode().find(".CommentNode-container").childAt(1).text()).toMatch(text)
    })


    /* Below button tests assume that current user is comment author */
    
    it('Delete button (Field comp.) receives correct props', () => {
      const props = commentNode().find({ type: 'linkButton', label: 'Delete' }).props()
      const expectedProps = {
        type: 'linkButton',
        label: 'Delete',
        inline: true,
        onClick: expect.any(Function)
      }
      expect(props).toStrictEqual(expectedProps)
    })
    
    it('deleteButton() func receives correct parms', () => {
      const postId = 'some_post_id123'
      props.postId = postId
      const commentId = 'some_comment_id123'
      props.comment._id = commentId

      props.deleteButton = jest.fn()
      commentNode().find({ type: 'linkButton', label: 'Delete' }).simulate('click')
      expect(props.deleteComment).toHaveBeenCalledWith(postId, commentId, props.fetchUsersStats)
    })
  })

  describe('mapStateToProps', () => {
    it('returns expected object', () => {
      const state = { user: { id: 'id123' } }
      expect(mapStateToProps(state)).toEqual({ authedUserId: 'id123' })
    })
  })

  it('must pass a snapshot test', () => {
    expect(commentNode()).toMatchSnapshot()
  })
})
