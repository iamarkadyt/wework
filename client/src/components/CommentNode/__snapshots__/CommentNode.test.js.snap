// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentNode must pass a snapshot test 1`] = `
<div
  className="CommentNode-container"
>
  <p>
    <b>
      John Grapard
    </b>
     said at 
    September 18, 2018, 10:24:07 PM
    :
  </p>
  <p
    className="CommentNode-container-Node-body"
  >
    Hahahahahahahaaa!
  </p>
  <div>
    <field
      inline={true}
      label="Delete"
      onClick={[Function]}
      type="linkButton"
    />
  </div>
</div>
`;
